pkg/controller/network/controller.go

```diff
@@ -0,0 +1,84 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package network
+
+import (
+	extensionspredicate "github.com/gardener/gardener-extensions/pkg/predicate"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// FinalizerName is the Network controller finalizer.
+	FinalizerName = "extensions.gardener.cloud/Network"
+	// ControllerName is the name of the controller.
+	ControllerName = "network"
+)
+
+// AddArgs are arguments for adding an Network controller to a manager.
+type AddArgs struct {
+	// Actuator is an Network actuator.
+	Actuator Actuator
+	// ControllerOptions are the controller options used for creating a controller.
+	// The options.Reconciler is always overridden with a reconciler created from the
+	// given actuator.
+	ControllerOptions controller.Options
+	// Predicates are the predicates to use.
+	// If unset, GenerationChangedPredicate will be used.
+	Predicates []predicate.Predicate
+}
+
+// DefaultPredicates returns the default predicates for a Network reconciler.
+func DefaultPredicates(typeName string, ignoreOperationAnnotation bool) []predicate.Predicate {
+	if ignoreOperationAnnotation {
+		return []predicate.Predicate{
+			extensionspredicate.HasType(typeName),
+			extensionspredicate.GenerationChanged(),
+		}
+	}
+
+	return []predicate.Predicate{
+		extensionspredicate.HasType(typeName),
+		extensionspredicate.Or(
+			extensionspredicate.HasOperationAnnotation(),
+			extensionspredicate.LastOperationNotSuccessful(),
+			extensionspredicate.IsDeleting(),
+		),
+	}
+}
+
+// Add creates a new network Controller and adds it to the Manager.
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, args AddArgs) error {
+	args.ControllerOptions.Reconciler = NewReconciler(mgr, args.Actuator)
+	return add(mgr, args.ControllerOptions, args.Predicates)
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, options controller.Options, predicates []predicate.Predicate) error {
+	ctrl, err := controller.New(ControllerName, mgr, options)
+	if err != nil {
+		return err
+	}
+
+	if err := ctrl.Watch(&source.Kind{Type: &extensionsv1alpha1.Network{}}, &handler.EnqueueRequestForObject{}, predicates...); err != nil {
+		return err
+	}
```

Comment:
We probably need also to have a mapper for `Cluster` resources due to the `ShootNotFailed` predicate.